### DISCUSSION TITLE
DataGrid not accessible via keyboard

### DISCUSSION THREAD

**Discussion by AArnott** [AvaloniaUI/Avalonia#10175] [Q&A] [answered]
**Describe the bug**

There seems to be no way to edit data in a DataGrid using the keyboard. 

A DataGrid with a textbox column and a checkbox column require **3 clicks** of the mouse just to set focus on the text box. I honestly cannot find any keystroke that sets focus on the textbox, making this control inaccessible to keyboard users.

**Expected behavior**

I expect spacebar, F2, enter, tab, and/or typing characters to be interpreted as interest in editing the selected row.

**Screenshots**

In the animation below, you can see the 3 click process for editing a textbox before the keyboard can be used to edit.
![0be7758c-0f11-44ba-94e6-4362b000e36f](https://user-images.githubusercontent.com/3548/216344762-d3333fd2-1a92-41e2-b0ab-457aaf6b4f8f.gif)

**Desktop (please complete the following information):**

- OS: Windows 11
- Version 0.10.18

**Comment by timunie** [comment]
At least `[F2]` works for me. Do you have a minimal sample for us?

**Comment by AArnott** [comment]
Thanks for the fast reply. In developing a minimal repro, I learned a bit about the problem. Actually the simplest case works great. It only takes two clicks to focus the textbox and F2 works as you say.

The problem begins when the CellEditingTemplate isn't *just* a TextBox, but a TextBox within a Grid. 
I use a Grid because (as you can see in my animation above) I want watermark text to appear as well. Can you please recommend a way to avoid the negative interaction changes that come from having more than just a TextBox?

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:AvaloniaApplication1.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="300" d:DesignHeight="250"
		Width="300" Height="250"
        x:Class="AvaloniaApplication1.Views.MainWindow"
        Icon="/Assets/avalonia-logo.ico"
        Title="AvaloniaApplication1">

	<Window.DataContext>
		<vm:MainWindowViewModel/>
	</Window.DataContext>

	<DataGrid Items="{Binding Rows}" AutoGenerateColumns="False" CanUserSortColumns="False">
		<DataGrid.Columns>
			<DataGridTemplateColumn Header="Name">
				<DataGridTemplateColumn.CellTemplate>
					<DataTemplate>
						<TextBlock Text="{Binding Name}" />
					</DataTemplate>
				</DataGridTemplateColumn.CellTemplate>
				<DataGridTemplateColumn.CellEditingTemplate>
					<DataTemplate>
						<Grid>
							<TextBox Text="{Binding Name}" />
						</Grid>
					</DataTemplate>
				</DataGridTemplateColumn.CellEditingTemplate>
			</DataGridTemplateColumn>
		</DataGrid.Columns>
	</DataGrid>
</Window>
```

The view model is:

```cs
using DynamicData;
using ReactiveUI;
using System.Collections.ObjectModel;
using System.Linq;

namespace AvaloniaApplication1.ViewModels;

public class MainWindowViewModel : ViewModelBase {
    public MainWindowViewModel()
    {
        this.Rows.AddRange(Enumerable.Range(1, 5).Select(i => new Person { Name = $"Name {i}" }));
    }

    public ObservableCollection<Person> Rows { get; } = new();
}

public class Person : ViewModelBase {
    private string name;
    public string Name {
        get => this.name;
        set => this.RaiseAndSetIfChanged(ref this.name, value);
    }
}
```

**Comment by timunie** [comment, marked as answer]
Hmmm several options:

1. Stick with TemplateColumn. Listen to `PrepareCellForEditing` event and manually set focus to the `TextBox`
2. Use `DataGridTextColumn` instead and add a `Style` for `TextBox` inside `DataGrid.Styles`. In this Style you can use `InnerRightContent` and add your `Watermark`
3. Add a `Style` for `TextBox`, but roll your own `ControlTemplate` which does what you need. 

Hope any of the suggestions makes sense. 


As you confirm that the issue is gone in a simpler scenario, let me convert this issue into a Q&A. I hope you are OK with that.

**Reply by AArnott** [reply]
I like option 2. But how would my style Selector point to the TextBox in the right column? The `DataGridTextColumn` class doesn't have a `Classes` property. For now, this happens to be the only text column in my grid so I'm not blocked, but I have another app where I'd like to use this option where I have many textbox columns.

**Reply by timunie** [reply]
this should work:

```xml
<DataGrid.Styles>
    <Style Selector="DataGridCell.watermark > TextBlock">
        <Setter Property="TextWrapping" Value="Wrap" />
    </Style>
    <Style Selector="DataGridCell.watermark > TextBox">
        <Setter Property="TextWrapping" Value="Wrap" />
        <Setter Property="InnerRightContent" Value="Hello World" />
    </Style>
</DataGrid.Styles>

<DataGridTextColumn CellStyleClasses="watermark" Binding="{Binding Text}" />
```
